src/sync/mpsc/mod.rs

```diff
@@ -0,0 +1,693 @@
+                    thread::yield_now();
+                }
+            }
+        }
+    }
+
+    // Try to park the receiver task
+    fn try_park(&self) -> TryPark {
+        let curr = self.inner.state.load(Ordering::SeqCst);
+        let state = decode_state(curr);
+
+        // If there are pending messages, then there is no need to park.
+        if state.num_messages > 0 {
+            return TryPark::NotEmpty;
+        }
+
+        // If the channel is closed, then there is no need to park.
+        if !state.is_open {
+            return TryPark::Closed;
+        }
+
+        // First, track the task in the `recv_task` slot
+        let mut task = self.inner.recv_task.lock().unwrap();
+        *task = Some(task::park());
+
+        // Ensure that there are still no messages. It is possible for a
+        // message to have been sent while updating the task slot.
+        let curr = self.inner.state.load(Ordering::SeqCst);
+        let state = decode_state(curr);
+
+        if state.num_messages == 0 {
+            TryPark::NotEmpty
+        } else {
+            TryPark::Parked
+        }
+    }
+
+    fn dec_num_messages(&self) {
+        let mut curr = self.inner.state.load(Ordering::SeqCst);
+
+        loop {
+            let mut state = decode_state(curr);
+
+            state.num_messages -= 1;
+
+            let next = encode_state(&state);
+            let actual = self.inner.state.compare_and_swap(curr, next, Ordering::SeqCst);
+
+            if actual == curr {
+                return;
+            }
+
+            curr = actual;
+        }
+    }
+}
+
+impl<T> Stream for Receiver<T> {
+    type Item = T;
+    type Error = ();
+
+    fn poll(&mut self) -> Poll<Option<T>, ()> {
+        loop {
+            // Try to read a message off of the message queue.
+            let msg = match self.next_message() {
+                Async::Ready(msg) => msg,
+                Async::NotReady => {
+                    // There are no messages to read, in this case, attempt to
+                    // park. The act of parking will verify that the channel is
+                    // still empty after the park operation has completed.
+                    match self.try_park() {
+                        TryPark::Parked => {
+                            // The task was parked, and the channel is still
+                            // empty, return NotReady.
+                            return Ok(Async::NotReady);
+                        }
+                        TryPark::Closed => {
+                            // The channel is closed, there will be no further
+                            // messages.
+                            return Ok(Async::Ready(None));
+                        }
+                        TryPark::NotEmpty => {
+                            // A message was sent while trying to park. Yield
+                            // the thread and try the poll operation again.
+                            thread::yield_now();
```

Comment:
Hm I think I may have missed this before. The `yield_now` calls above for working with `Inconsistent` I'm already slightly uncomfortable with. I feel ok with those though in the sense that there's literally a sequence of two instructions where when preempted in the middle will cause `Inconsistent`.

Here, though, there's perhaps an arbitrarily large amount of time between the increment of the number of messages and when the data's actually sent, causing a receiver to spin here for awhile. Notably in this window a call to `task::park` is made which could involve allocating memory, involving syscalls, etc, etc.

Is there perhaps an alternative to resolve this race?